src/storage/core.rs

```diff
@@ -415,100 +417,88 @@ where
+            } else if data.is_deleted() {
+                if all_entries.is_empty() {
+                    return Ok(data);
+                } else {
+                    return Ok(ReadResult::Found(all_entries));
+                }
+            }
+        }
+        debug!(
+            "storage core read from non-active total {} entries",
+            all_entries.len()
+        );
+        debug_assert!(all_entries
+            .iter()
+            .zip(all_entries.iter().skip(1))
+            .all(|(x, y)| x.header().created() >= y.header().created()));
+        Ok(ReadResult::Found(all_entries))
     }
 
-    async fn read_with_optional_meta(&self, key: &K, meta: Option<&Meta>) -> Result<Bytes> {
+    async fn read_with_optional_meta(
+        &self,
+        key: &K,
+        meta: Option<&Meta>,
+    ) -> Result<ReadResult<Bytes>> {
         debug!("storage read with optional meta {:?}, {:?}", key, meta);
         let safe = self.inner.safe.read().await;
         if let Some(ablob) = safe.active_blob.as_ref() {
             match ablob.read().await.read_any(key, meta, true).await {
                 Ok(data) => {
-                    debug!("storage read with optional meta active blob returned data");
-                    return Ok(data);
+                    if data.is_presented() {
+                        debug!("storage read with optional meta active blob returned data");
+                        return Ok(data);
+                    }
                 }
                 Err(e) => debug!("read with optional meta active blob returned: {:#?}", e),
             }
         }
         Self::get_any_data(&safe, key, meta).await
     }
 
-    async fn get_data_last(safe: &Safe<K>, key: &K, meta: Option<&Meta>) -> Result<Bytes> {
-        let blobs = safe.blobs.read().await;
-        let possible_blobs = blobs
-            .iter_possible_childs_rev(key)
-            .map(|(id, blob)| async move {
-                if !matches!(blob.data.check_filter(key).await, FilterResult::NotContains) {
-                    Some(id)
-                } else {
-                    None
-                }
-            })
-            .collect::<FuturesOrdered<_>>()
-            .filter_map(|x| x)
-            .collect::<Vec<_>>()
-            .await;
-        debug!(
-            "len of possible blobs: {} (start len: {})",
-            possible_blobs.len(),
-            blobs.len()
-        );
-        let stream: FuturesOrdered<_> = possible_blobs
-            .into_iter()
-            .filter_map(|id| blobs.get_child(id))
-            .map(|blob| blob.data.read_any(key, meta, false))
-            .collect();
-        debug!("read with optional meta {} closed blobs", stream.len());
-        let mut task = stream.skip_while(Result::is_err);
-        task.next().await.ok_or_else(Error::not_found)?
-    }
-
-    #[allow(dead_code)]
-    async fn get_data_any(safe: &Safe<K>, key: &K, meta: Option<&Meta>) -> Result<Bytes> {
+    async fn get_any_data(
```

Comment:
`get_any_data` is not needed and can be removed. You probably removed the wrong function